### ISSUE TITLE
RIP

### ISSUE THREAD

**Issue by hazre** [AdGoBye/AdGoBye#105]
![image](https://github.com/user-attachments/assets/7164e870-bd29-4251-9774-a3d6c62f4a93)
<https://creators.vrchat.com/releases/#changes-in-372-beta2>

*cough* time to look to the alternative way *cough*

**Comment by RaspberryKitty1** [comment]
might mark the end of this project rip...

**Comment by regalialong** [comment]
This should pretty much only affect scripting I believe, Oct 3rd dev blog speaks in more depth about the protections themselves and upstream explicitly mentions that it's difficult to protect the bundle itself as a whole.

And in all honesty, I think this is probably for the best because it takes a big nuanced argument off the table that AGB can enable abuse per script injection.

They said they did add a CRC32 check for beta2 though, which means I'll need to cook up something for that.

**Comment by hazre** [comment]
I haven't verified this but my friend told me how it works is they calculate a world bundle hash on the server and that they compare it to the local cached bundle hash thats computed on your PC before the world loads.

**Comment by kitlith** [comment]
I did some research back at the end of november/end of december before getting distracted, thought about this again recently, so I figured I'd write up my notes and what I remember here:

If this is unity's built in CRC check for asset bundles (which the dev post seemed to imply), I did a little bit of research into how that's computed. If we see the message `"CRC Mismatch. Provided %x, calculated %x from data. Will not load AssetBundle \'%s\'"` inside of logs somewhere, this is almost certainly what VRChat is using. (I never actually checked >_>)

There's a piece of documentation that I copied (but I don't remember from where) that says:
> When provided, the loading system calculates the checksum of the uncompressed content of the AssetBundle before loading it.

The CRC in question appears to be crc32-adler, according to the CRC table I found. This is handled with native code inside of UnityPlayer.dll, inside `ArchiveStorageCreator`, and the CRC check itself is performed within `AssetBundleLoadFromStreamAsyncOperation::FinalizeArchiveCreator` (or similar methods.) I did my research on a non-il2cpp version of the engine, bundled with the editor.

A diagram I made tracing the calls that feed the CRC:

![Image](https://github.com/user-attachments/assets/a48734a5-85c5-405e-a765-1f97c00ea337)

(and of course, once we can calculate the CRC ourselves, we can do linear algebra to manipulate the CRC back to its original value.)

A good way to test your CRC calculation code might be to find a game that uses unity's addressables system (was beat saber one? i can't remember.) and compare your calculated crc to the one stored in the catalog file. There exist some tools used for modding to set every crc in a catalog file to 0 to bypass the crc check, heh,

<details>
<summary>An aside on Udon Code Signing</summary>
As of 11/30/2024, it looked like the SDK on the creator's machine was generating the private key and signing the udon scripts. (Build & Test/Upload both computed the signatures.)

![Image](https://github.com/user-attachments/assets/37efbdca-3266-4577-af73-10e20f0f5791)

Technically, a world author could extract the private key and modify the scripts of their own world, share the private key for others to be able to mod the scripts in their world, or use a known static key.

VRChat has the power at any time (and may already be doing, I didn't check) to replace the key being used and recompute the signatures on their end, if people start doing this and they don't like it. As-is, though, it's basically up to the world creator who can modify/add udon scripts to their worlds.
</details>

I hope someone finds this helpful, as I'm not sure if/when I'm going to swing back around to doing further research on this.

**Comment by regalialong** [comment]
Thanks for the CRC writeup, I'm currently stuck between projects and recently also employment.

I don't think we get that explicitly info about CRC failures, the closest is this when appending junk to the file in the logs:
> Error      -  Error Downloading or Instantiating Custom Scene - YTS 2․1 - YouTube Search‚ Subtitles‚ Quest - 12 - Error loading asset bundle from ****: Unable to write data

The file version gets deleted and the user gets brought back to their home world.
I'll have to see when I find the time (and admittedly the motivation) to fix it, obviously PRs are more than welcome if someone beats me to it.

I believe also the cache structure changed the last time AGB was functional? Pretty sure we can use VRCX code for that :P 
At the very least the indexer probably need tweaks to handle duplicated stuff, since different content variants are becoming more common ("security" variants, imposters, etc)

**Comment by regalialong** [comment]
For what it's worth, Upstream explains in Dev Update 3 October 2024 that they do just employ the stock CRC check and they identify it's really meant corruption checking and not for integrity checking (like the script signatures are)

**Comment by regalialong** [comment]
Oh wait, I'm stupid, spotted the CRC:
>  [AssetBundleDownloadManager] Verifying world bundle load with CRC ****FE62

**Comment by kitlith** [comment]
> For what it's worth, Upstream explains in Dev Update 3 October 2024 that they do just employ the stock CRC check

yeah that's what the "(which the dev post seemed to imply)" was meant to refer to. I'm using uncertain terms because of time (it's been months since i looked at this!) and because I never actually checked what the vrchat client was doing -- just made inferences from dev updates and unity documentation.

> and they identify it's really meant corruption checking and not for integrity checking (like the script signatures are)

which is why manipulating the CRC back to its original value is possible, yeah. it feels kinda like the elephant in the room: VRChat's aware of it, you're aware of it, it's just waiting for someone to assemble the details to take advantage of it and ride the elephant off into the sunset. At least, until VRChat decides to do a full integrity check and shut this approach down.

---

random thing i just remembered wasn't part of my initial infodump: the CRC I was looking at doesn't seem to depend on the header of the asset bundle at all.

**Comment by regalialong** [comment]
Yeah, this is very much a 'we have the tools but someone needs to assemble them' situation


> random thing i just remembered wasn't part of my initial infodump: the CRC I was looking at doesn't seem to depend on the header of the asset bundle at all.

This can't be a CRC of the entire bundle, iirc they bumped engine versions locally of incompatible bundles so they can be loaded by newer clients
I believe this is partially also why they mentioned that making a strong signature of the entire bundle is a challenge

I've been spending a few hours trying to recreate the CRC but haven't gotten anywhere

**Comment by kitlith** [comment]
I'm considering writing a mod that'll hook FeedCRC and just log the data that was passed to it, so that you can get a log for an asset bundle of "what data was CRC'd when" and compare to your own implementation.

Do you have a preferred game/mod loader I should try to target for this?

**Comment by regalialong** [comment]
You're starting to tread some territories that probably shouldn't be discussed on here, as much as I appreciate the depth you're willing to go for this :P
I did notice you joined the Matrix so feel free to hit me up there

As much as I hate pulling out this project, https://github.com/AXiX-official/Unity_CRC32_Bypass works but I haven't managed to recreate this within AssetsTools.NET since they essentially remade AT from scratch (likely for the sake of UnityCN?)

**Comment by kitlith** [comment]
Alright, we've made some progress together. I threw together a [hackjob](https://github.com/kitlith/UnityAsset.NET/commit/c4462be7d3cc1695efcb879da88ca382a73aa401) to aid in debugging the CRC implementation for AssetsTools. Good news, we can just use DataStream, and the way it was being fed into the CRC implementation just needed to be tweaked.

Currently, we're stuck on adding the collision bytes to the stream. In theory, we can probably just add the collision bytes [here](https://github.com/nesrak1/AssetsTools.NET/blob/2207c788da2dc9198fb1a91ed67b212f1c1cad2d/AssetTools.NET/Standard/AssetsBundleFileFormat/AssetBundleFile.cs#L182) when writing the bundle back out to the disk. The problem is that we don't want to fork if we don't need to. (I think I distracted @regalialong before they tested it though)

We may be able to handle this while staying within the bounds of AssetTools by adding an extra file, or asset, or field, or something. The potential challenge is either in ensuring that we add *exactly* 4 bytes to the end of the file, or that we know exactly *where* in the data stream the 4 bytes are placed. (The former we already have code for, the latter would require something probably a little less fancy than https://github.com/resilar/crchack, since we can probably rely on all 4 bytes being contiguous. or maybe we'll be replacing the last 4 bytes instead of adding them back in afterwards.)

**Comment by regalialong** [comment]
Ooooooookay, figured I'd deliver the bad news. Latest patch (1607) has turned bundles highly entropic, even after decompression.

The build includes `Safety & Security changes.` which require a cache reset. Considering upstream would shout from every rooftop if they'd made a new compression scheme, pretty sure the writing on the wall is clear; ***bundles are encrypted now.***

I don't believe they're encrypted server-side (since that'd be very costly for something the client will eventually decrypt anyway). (Technical) readers should be able to spot what's going on and what would be the plan forward.

I'll be fully transparent, I don't think we can deal with this without stepping on someone's toes.
I took the scripting signatures as a signal that they're not interested in protecting the whole bundle, only the scripting which by far would the be the easiest to exploit in a malicious manner, but I guess I was wrong on that.

Besides, even if we end up decrypting bundles (which depending on the methods could reasonably ***endanger users and maintainers***), it's obvious from my side that I've not had the time and / or motivation to contribute to this.

But hey, everything is MIT licensed, I make no possessions.
'Twas fun, but it should be obvious for everyone involved that this was bound to come to an end somewhere.

I'll not `So long and thanks for all the fish` you all just yet, but this is probably end of the road.

**Comment by kitlith** [comment]
Grain of salt, I haven't done any technical research on this yet -- this is just my thoughts in response to the news.

> I took the scripting signatures as a signal that they're not interested in protecting the whole bundle, only the scripting which by far would the be the easiest to exploit in a malicious manner, but I guess I was wrong on that.

I've been expecting this for some time, given that they have user experience incentives to make ripping as difficult as possible. I vaguely recall reading *something* that made it clear they were working on this. Not that much of a stretch for them to apply it to worlds as well as avatars.

The scripting stuff is still useful to them as a defence-in-depth technique, though. (Even if someone is still able to decrypt/encrypt bundles, there's only so much they can do to udon scripts.)

> I don't believe they're encrypted server-side (since that'd be very costly for something the client will eventually decrypt anyway).

For what it's worth, I think it's totally possible for them to do server side encryption. There's been a server-side bundle variant generation system for a little bit which processes your avatar on first load if the needed variant hasn't been generated yet. I suspect it wouldn't be too much extra computation on their end to encrypt things afterwards when doing that processing and store the encrypted bundle. Then "all" they need to do is ensure that the decryption key is never stored on disk (for per-avatar/world decryption keys) and is always served over the network when needed, or use a white-box decryption implementation for a shared decryption key.

EAC making it difficult to inspect/hook the game is useful (to vrchat) here in preventing someone from just logging the decryption keys directly from the game. It gives this scheme more teeth than ChilloutVR (which at one point had a mod re-implementing asset bundle decryption to make it faster because the original method was slow), though I suspect there may still be a workaround that doesn't require breaking EAC. For now. However, either way, that would still be stepping on toes.

> I don't think we can deal with this without stepping on someone's toes.

> I'll not `So long and thanks for all the fish` you all just yet, but this is probably end of the road.

Agreed, assuming they've implemented this in the way I'm visualizing (or better). I'll wait for someone (or myself) to do a deep dive to figure out exactly what's going on before calling it for sure, but it's not looking good.

**Comment by regalialong** [comment]
Yeah, your conjecture should be correct in this regard.

> though I suspect there may still be a workaround that doesn't require breaking EAC.

I said this in LVRA, it's not as much about the method (the client has to decrypt it at some point, there's no silver bullet against someone who controls the hardware) as much as defeating the encryption will run us afoul of DMCA anti-circumvention. You can justify that a CRC32 is not meant as a mechanism to prevent access to bundles, but encryption is.

I've been explicit in the past that [users should keep their hands off avatars](https://github.com/AdGoBye/AdGoBye/blob/main/CONTRIBUTING.md#couldnt-this-code-disable-avatar-things-as-well) anyway, so avatars are out of scope for this project. We're probably in the crossfire of ripping protections, which would be fine, but I don't think there's an interest for ripping worlds, though maybe upstream knows a thing I don't.

Even if the method fell off a truck and into an AGB PR, I'd rather just call it quits here than play that kind of game and endanger everyone collectively.

**Comment by Bluscream** [comment]
and so AdGoBye goes bye 🙏🏻 until the next nerd gets annoyed enough to figure smth out
